tempesta_fw/classifier/frang.c

```diff
@@ -993,6 +1137,100 @@ frang_free_ct_vals(TfwCfgSpec *cs)
 }
 
 static int
+frang_register_fsm_resp(void)
+{
+	int r = tfw_gfsm_register_fsm(TFW_FSM_FRANG_RESP, frang_resp_handler);
+	if (r) {
+		TFW_ERR("\nfrang: can't register response fsm");
+		return r;
+	}
+
+	fsm_hook_resp_prio = tfw_gfsm_register_hook(TFW_FSM_HTTP,
+						    TFW_GFSM_HOOK_PRIORITY_ANY,
+						    TFW_HTTP_FSM_RESP_MSG_FWD,
+						    TFW_FSM_FRANG_RESP,
+						    TFW_FRANG_RESP_FSM_INIT);
+	if (fsm_hook_resp_prio < 0) {
+		TFW_ERR("\nfrang: can't register gfsm msg fwd hook");
+		tfw_gfsm_unregister_fsm(TFW_FSM_FRANG_RESP);
+		return fsm_hook_resp_prio;
+	}
+
+	return 0;
+}
+
+static int
+frang_parse_ushort(const char *s, unsigned short *out)
+{
+	int n;
+	if(tfw_cfg_parse_int(s, &n) || n < 1 || n > USHRT_MAX) {
+		TFW_ERR_NL("frang: http_resp_code_block: \"%s\" isn't a valid value\n", s);
+		return -EINVAL;
+	}
+	*out = n;
+	return 0;
+}
+
+/**
+ * Save response code block configuration
+ */
+static int
+frang_set_rsp_code_block(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int n, i;
+	FrangHttpRespCodeBlock *cb;
+
+	if (ce->val_n < 3) {
+		TFW_ERR_NL("frang: http_resp_code_block: too few arguments\n");
+		return -EINVAL;
+	}
+
+	cb = kmalloc(sizeof(FrangHttpRespCodeBlock), GFP_KERNEL);
+	if (!cb)
+		return -ENOMEM;
+	((FrangCfg *)cs->dest)->http_resp_code_block = cb;
+
+	i = ce->val_n - 2;
+	memset(cb->codes, 0, sizeof(cb->codes));
+	while (--i >= 0) {
+		if (tfw_cfg_parse_int(ce->vals[i], &n) || !frang_resp_code_range(n)) {
+			TFW_ERR_NL("\nfrang: http_resp_code_block: invalid HTTP error code \"%s\"", ce->vals[i]);
+			return -EINVAL;
+		}
+		/*
+		 * Atomic restriction isn't needed here
+		 */
+		__set_bit(FRANG_HTTP_CODE_BIT_NUM(n), cb->codes);
+	}
+
+	if (frang_parse_ushort(ce->vals[ce->val_n - 2], &cb->limit)
+	    || frang_parse_ushort(ce->vals[ce->val_n - 1], &cb->tf))
+		return -EINVAL;
+
+	cb->sf = cb->tf / FRANG_RESP_CODE_TIME_FRAME
+		+ (cb->tf % FRANG_RESP_CODE_TIME_FRAME != 0);
+	/*
+	 * (cb->tf - 1) / cb->sf is a maximum value of ai in frang_resp_handler()
+	 */
+	cb->stat_buf_len = ((cb->tf - 1) / cb->sf) + 1;
+
+
+	return frang_register_fsm_resp();
```

Comment:
The good point in the current code is that we don't register the GFSM hook if the feature wasn't configured and we don't need to use the hook. It's better for runtime. Meantime, `frang_register_fsm_resp()` prints its own error messages, so the user sees the reason for the failure, but, yes, only at configuration phase.

Reply:
I meant call `frang_register_fsm_resp()` in `frang_init()` only if `frang_cfg.http_resp_code_block` is non-null. For me it looks that `frang_init()` is the best place to register GFSM, but i don't insist.

Reply:
> I meant call frang_register_fsm_resp() in frang_init() only if frang_cfg.http_resp_code_block is non-null. 

Can we check the configuration option already in `frang_init()`? If so, then it will be the best solution. I supposed that `frang_init()` is called on the module loading, before reading the configuration file, so we don't know whether the option is switched on or off.

Reply:
Sorry, both you guys, for the some reason i've messed up everything altogether: both init and start of Tempesta's module. I was wrong in this case, nothing is wrong here.